z3/src/z3_solver.cpp

```diff
@@ -0,0 +1,844 @@
+};
+void Z3Solver::add_selector(DatatypeConstructorDecl & dt,
+                            const std::string & name,
+                            const Sort & s) const
+{
+  throw NotImplementedException("Z3Solver::add_selector");
+};
+void Z3Solver::add_selector_self(DatatypeConstructorDecl & dt,
+                                 const std::string & name) const
+{
+  throw NotImplementedException("Z3Solver::add_selector_self");
+};
+
+Term Z3Solver::get_constructor(const Sort & s, std::string name) const
+{
+  throw NotImplementedException("Z3Solver::get_constructor");
+};
+Term Z3Solver::get_tester(const Sort & s, std::string name) const
+{
+  throw NotImplementedException("Z3Solver::get_testeer");
+};
+
+Term Z3Solver::get_selector(const Sort & s,
+                            std::string con,
+                            std::string name) const
+{
+  throw NotImplementedException("Z3Solver::get_selector");
+};
+
+Term Z3Solver::make_term(int64_t i, const Sort & sort) const
+{
+  SortKind sk = sort->get_sort_kind();
+  expr z_term = expr(ctx);
+  if (sk == INT)
+  {
+    z_term = ctx.int_val(i);
+  }
+  else if (sk == REAL)
+  {
+    z_term = ctx.real_val(i);
+  }
+  else if (sk == BV)
+  {
+    z_term = ctx.bv_val(sort->get_width(), i);
+  }
+  else
+  {
+    string msg("Can't create value ");
+    msg += i;
+    msg += " with sort ";
+    msg += sort->to_string();
+    throw IncorrectUsageException(msg);
+  }
+
+  return std::make_shared<Z3Term>(z_term, ctx);
+}
+
+Term Z3Solver::make_term(const std::string val,
+                         const Sort & sort,
+                         uint64_t base) const
+{
+  expr z_term = expr(ctx);
+  SortKind sk = sort->get_sort_kind();
+
+  if (base != 10)
+  {
+    throw NotImplementedException("Does not support base not equal to 10.");
+  }
+
+  if (sk == BV)
+  {
+    z_term = ctx.bv_const(val.c_str(), sort->get_width());
+  }
+  else if (sk == REAL)
+  {
+    z_term = ctx.real_const(val.c_str());
+  }
+  else if (sk == INT)
+  {
+    z_term = ctx.int_const(val.c_str());
+  }
+  else
+  {
+    string msg("Can't create value ");
+    msg += val;
+    msg += " with sort ";
+    msg += sort->to_string();
+    throw IncorrectUsageException(msg);
+  }
+
+  return std::make_shared<Z3Term>(z_term, ctx);
+}
+
+Term Z3Solver::make_term(const Term & val, const Sort & sort) const
+{
+  std::shared_ptr<Z3Term> zterm = std::static_pointer_cast<Z3Term>(val);
+  std::shared_ptr<Z3Sort> zsort = std::static_pointer_cast<Z3Sort>(sort);
+
+  if (zsort->is_function || zterm->is_function)
+  {
+    throw IncorrectUsageException(
+        "Cannot create constant array with function element");
+  }
+
+  Z3_ast c_array = Z3_mk_const_array(ctx, zsort->type, zterm->term);
+  expr final = to_expr(ctx, c_array);
+  return std::make_shared<Z3Term>(final, ctx);
+}
+
+void Z3Solver::assert_formula(const Term & t)
+{
+  throw NotImplementedException(
+      "Term iteration not implemented for Z3 backend.");
```

Comment:
Just FYI, looks like this exception message was copy-pasted. This one doesn't need to be fixed now since it will be implemented soon. But for anything that stays throwing an exception permanently, please update the message.